File: app/src/main/java/com/johnpetitto/orachat/ui/chats/ChatsFragment.java
package com.johnpetitto.orachat.ui.chats;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.arlib.floatingsearchview.FloatingSearchView;
import com.johnpetitto.orachat.OraChatApplication;
import com.johnpetitto.orachat.R;
import com.johnpetitto.orachat.data.chat.Chat;
import com.johnpetitto.orachat.data.chat.ChatModel;
import com.johnpetitto.orachat.ui.PagingScrollListener;
import com.johnpetitto.orachat.ui.chatroom.ChatroomActivity;

import java.util.List;

import butterknife.BindDimen;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

public class ChatsFragment extends Fragment implements ChatsView,
        FloatingSearchView.OnQueryChangeListener, ChatsAdapter.OnChatClickListener,
        PagingScrollListener.OnPageListener {

    @BindView(R.id.chats_recycler_view) RecyclerView recyclerView;
    @BindView(R.id.chats_search_view) FloatingSearchView searchView;
    @BindView(R.id.chats_progress_bar) ProgressBar progressBar;

    @BindDimen(R.dimen.dialog_margin) int dialogMargin;

    private Unbinder unbinder;

    private ChatsPresenter presenter;
    private ChatsAdapter adapter;
    private PagingScrollListener scrollListener;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_chats, container, false);
        unbinder = ButterKnife.bind(this, rootView);

        ChatModel model = ((OraChatApplication) getContext().getApplicationContext()).getChatModel();
        presenter = new ChatsPresenter(this, model);

        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(adapter = new ChatsAdapter(this));
        recyclerView.addOnScrollListener(scrollListener = new PagingScrollListener(this));

        searchView.setOnQueryChangeListener(this);

        presenter.getAllChatsByDate();

        return rootView;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        presenter.onDestroy();
        unbinder.unbind();
    }

    @OnClick(R.id.chats_new)
    public void newChat(View view) {
        Intent intent = new Intent(getActivity(), ChatroomActivity.class);
        intent.putExtra("chat_name", getString(R.string.new_chat));
        startActivity(intent);
    }

    @Override
    public void showLoading(boolean show) {
        progressBar.setVisibility(show ? View.VISIBLE : View.GONE);
        recyclerView.setVisibility(show ? View.INVISIBLE : View.VISIBLE);
    }

    @Override
    public void showError() {
        Toast.makeText(getContext(), R.string.generic_error, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void displayChatsByDate(List<Object> chatsGroupedByDate) {
        adapter.setChatsGroupedByDate(chatsGroupedByDate);
    }

    @Override
    public void showMoreChatsByDate(List<Object> chatsGroupedByDate) {
        adapter.addChatsGroupedByDate(chatsGroupedByDate);
    }

    @Override
    public void pageLoaded() {
        scrollListener.loadingComplete();
    }

    @Override
    public void onSearchTextChanged(String oldQuery, String newQuery) {
        presenter.searchChatsByName(newQuery); // TODO: throttle
    }

    @Override
    public void onChatClick(Chat chat) {
        Intent intent = new Intent(getActivity(), ChatroomActivity.class);
        intent.putExtra("chat_id", chat.getId());
        intent.putExtra("chat_name", chat.getName());
        startActivity(intent);
    }

    @Override
    public void loadPage() {
        presenter.loadMoreChats();
    }
}

File: app/src/main/java/com/johnpetitto/orachat/ui/chats/ChatsFragment.kt
package com.johnpetitto.orachat.ui.chats

import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.johnpetitto.orachat.OraChatApplication
import com.johnpetitto.orachat.R
import com.johnpetitto.orachat.show
import com.johnpetitto.orachat.ui.PagingScrollListener
import com.johnpetitto.orachat.ui.chatroom.ChatroomActivity
import kotlinx.android.synthetic.main.fragment_chats.*

class ChatsFragment: Fragment(), ChatsView, PagingScrollListener.OnPageListener {
    private lateinit var presenter: ChatsPresenter
    private val scrollListener = PagingScrollListener(this)
    private val chatsAdapter = ChatsAdapter { chat ->
        startActivity(Intent(activity, ChatroomActivity::class.java).apply {
            putExtra("chat_id", chat.id)
            putExtra("chat_name", chat.name)
        })
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater!!.inflate(R.layout.fragment_chats, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        recyclerView.apply {
            layoutManager = LinearLayoutManager(context)
            adapter = chatsAdapter
            addOnScrollListener(scrollListener)
        }

        search.setOnQueryChangeListener { _, newQuery -> presenter.searchChatsByName(newQuery) }

        newChat.setOnClickListener {
            startActivity(Intent(activity, ChatroomActivity::class.java).apply {
                putExtra("chat_name", getString(R.string.new_chat))
            })
        }

        val model = (context.applicationContext as OraChatApplication).chatModel
        presenter = ChatsPresenter(this, model)
        presenter.getAllChatsByDate()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        presenter.destroy()
    }

    override fun showLoading(show: Boolean) {
        progress.show(show)
        recyclerView.show(!show)
    }

    override fun showError() {
        Toast.makeText(context, R.string.generic_error, Toast.LENGTH_SHORT).show()
    }

    override fun displayChatData(chatData: List<Any>) = chatsAdapter.setChatData(chatData)
    override fun showMoreChatData(chatData: List<Any>) = chatsAdapter.addChatData(chatData)

    override fun pageLoaded() = scrollListener.loadingComplete()
    override fun loadPage() = presenter.loadMoreChats()
}
